README.md
[![License](https://img.shields.io/badge/license-GPL%20%28%3E=%203%29-lightgrey.svg?style=flat)](http://www.gnu.org/licenses/gpl-3.0.html)
<img align="right" src="cenith.png" alt="drawing" width="350">


## Installation
Install via devtools from github:

``` r
devtools::install_github("SchoenbergA/CENITH@master",build_vignettes = TRUE)
```


NEWS.md
# CENITH 0.1.2
patch version

* bugfixes
With a new rdgal version now the strings for the vector example data contain additional +towgs84=0,0,0,0,0,0,0.
With those differnet strings (even if projcetion is still teh same) some functions and the vignette would not work.
Fixed by setting the crs of the vectors to crs of rasters by hand in example and vignette code.

* simplified example data loading

* update vignette
vignettes now come in hmtl format for direct download via github.
Install without "buildvignette=T" is now recommended.

# CENITH 0.1.1
patch version

* bugfixes general
For all functions using 'TreeSeg' - In TreeSeg now the window diameter is not set to a maximum. This leads to no more errors is the windiameter is wider.
Correct name for CENITH-Package.R in help is used.
Set depending packages in description.

* bugfixes and updates
TreeSeg - default MaxWinDiameter in subfucntion is now set to NULL. 
TreeSeg - corrected example added. Added require() to example.
BestSegVal - now 3rd try in ETA check works. Further now continues even if all 3 trys do not work.
BestSegVal - spelling corrections and added some cat() codes.
BestSegVal - add require() to example.
TreeSegCV - corrections in help
TreeSegCV - spelling corrections

# CENITH 0.1.0
* release version for GitHub (not rdy for CRAN)

# CNEITH 0.0.99.5
* spelling corrections

correction of spelling for all functions.
some added links to other functions.


# CENITH 0.0.99.4
* bugfixes

fixed TreeSegCV - now if a combination leads to an error the error is printen correctly and the function continues.

# CENITH 0.0.99.3
* new features
update vignette - add result='hide' to chunks for increased overview in tutorial.

# CENITH 0.0.99.2

* new features
update vignette - add part 2 (BestSegVal) and 3 (TreeSegCV) to vignette
add example data - chm and rgb for the AOI

# CENITH 0.0.99.1

* new features
add vignette - tutorial part 1 treseg 

# CENITH 0.0.99.0
Prerelease version 0.99

* new features
CENITH - package - add package script with imports and descriptions
add Example Data for Crossvalidation

* bugfixes
some major changes in BestSegVal cat codes.

# CENITH 0.0.0.97
* new features
complete clean up of Development functions (keep stable version of BestSegVal)

TreeSegVal - now iterates over a,b,h,MIN, and chm filters. Advanced ETA estimation added (now uses full iterations of TreeSeg). Further cleand up "cat" codes to more easy see the ETA.

Added all new Example data.

# CENITH 0.0.0.96
* new features
TreeSegValEXPRMTL - advanced version of TreeSegVal which additionally iterates over MIN and supports filtering for chm.

# CENITH 0.0.0.95
* bugfixes
TreeSegVal returns correct values for "Segment quality"

* new features
TreeSegCV - function to perform an n fold cross validation. Uses values (estimated with TreeSegVal recommended) to test the quality on other sides to receive a quality estimation of the inout values for the Segmentation in the AOI.

# CENITH 0.0.0.94
Update for MIN and MAX

* added features
TreeSeg -  Now returns Error, if after clipping no Segments are left. This is used to handle MIN and MAX in TreeSegVal.
TreeSegVal - now uses MIN and MAX in TreeSeg
           - improved results, including hit, oversegmentation and undersegmentation rates and absolut values.

# Cenith 0.0.0.93
Update

* updated features
BestSegValBETA - deleted older check for to high values. To estimate the time to target now up to 3 random iterations are used to sto time. if all 3 fail the function stops. added example with 3/4 iterations to fail.
*TreeSeg -  added MIN MAX from BestSegValBETA to basic Segmentation function. Crops polygons < MIN / >MAX

# CENITH 0.0.0.92
* development version for Validation of Segments

* new feature
BestSegValBETA - development verison contains error catching

# CENITH 0.0.0.91

* changes for development
TreeSeg - now returns Treepos + segments
BestSegVal - now computes more and other validation results

# CENITH 0.0.0.9000
initial version

*new feature

TreeSeg - function for TreeCrown Segmentation
BestSegVal - function to iterate over several a,b,h values to detect best fitting parameters for TreeSeg


R/TreeSeg.R
#' Compute TreeCrown Segments
#' @description computes polygon segments for TreeCrowns based on watershed algorithm
#' @param chm raster -  Canopy Height Model RasterLayer.
#' @param a numeric - single value for MovingWindow.
#' @param b numeric - single value for MovingWindow.
#' @param h numeric - maximum height of trees (in meter) to detect trees.
#' @param MIN numeric - the minimum area for crowns. Smaller polygons are cropped.
#' @param MAX numeric - the maximum area for crowns. Larger polygons are cropped.
#' @param CHMfilter numeric - uses a sum filter on the chm with a MovingWindow of (x*x), which must be odd. Default=1 no filter.
#' @param silent bolean - if TRUE the function will not print any progress messages (default=FALSE).
#' @return returns a PolygonLayer with segments
#' @details uses a MovingWindow of x*a+b to detect local maxima in a chm to compute TreeCrown Segments
#' * parameter selection - use \code{\link{BestSegVal}} to automated detect best fitting parameters for a, b, h, MIN and filter.
#' @note A 'brute force' segmentation with random parameters is not recommended. TreeSeg is mainly used to compute segments AFTER the validation of best fitting parameters with \code{\link{BestSegVal}}.
#' Further to estimate the quality of the computed polygons it is recommended to use \code{\link{TreeSegCV}} for a x-fold CrossValdiation over x different subareas. For full workflow see the 'CENITH_Tutorial'.
#' @seealso \code{\link{BestSegVal}},\code{\link{TreeSegCV}}
#' @author Andreas Schönberg
#' @examples
#' # load packages
#' require(CENITH)
#' require(raster)
#' require(mapview)
#' # load data
#' chm <- raster::raster(system.file("extdata","lau_chm.tif",package = "CENITH"))
#' # take a look on the data
#' plot(chm)
#' # NOTE: the example should NOT show to get optimal results (for this see 'BestSegVal')
#' # start segmentation
#' x <- TreeSeg(chm,a=0.3,b=0.5,h=1)
#' length(x)# amount of trees
#' # compare result with chm
#' mapview(chm)+x
#' # clip min and or max polygons
#' y <-TreeSeg(chm,a=0.2,b=0.5,h=2,MIN=10,MAX=100)
#' length(y)# amount of trees
#' mapview(chm)+y
#' @export TreeSeg
#' @aliases TreeSeg

TreeSeg <- function(chm=NULL,a,b,h,MIN=0,MAX=1000,CHMfilter=1,silent=FALSE){
  # function ForestTools vwf cleaned from cat code #############################
  vwf_clean <-function (CHM, winFun, minHeight = NULL, maxWinDiameter = NULL,
                              minWinNeib = "queen", verbose = FALSE)
  {
    if (verbose)
      if (!minWinNeib %in% c("queen", "rook"))
        stop("Invalid input for 'minWinNeib'. Set to 'queen' or 'rook'")
    CHM.crs <- as.character(raster::crs(CHM))
    CHM.prj <- regmatches(CHM.crs, regexpr("(?<=proj=).*?(?=\\s)",
                                           CHM.crs, perl = TRUE))
    if (length(CHM.prj) > 0 && CHM.prj %in% c(c("latlong", "latlon",
                                                "longlat", "lonlat"))) {
      warning("'CHM' map units are in degrees. Ensure that 'winFun' outputs values in this unit.")
    }
    roundRes <- round(raster::res(CHM), 5)
    if (roundRes[1] != roundRes[2])
      stop("Input 'CHM' does not have square cells")
    if (roundRes[1] == 0)
      stop("The map units of the 'CHM' are too small")
    if (!is.null(minHeight) && minHeight <= 0)
      stop("Minimum canopy height must be set to a positive value.")
    CHM.rng <- suppressWarnings(raster::cellStats(CHM, range))
    names(CHM.rng) <- c("min", "max")
    if (is.infinite(CHM.rng["max"]) | is.infinite(CHM.rng["min"])) {
      stop("Input 'CHM' does not contain any usable values. Check input data.")
    }
    if (!is.null(minHeight)) {
      if (minHeight >= CHM.rng["max"])
        stop("'minHeight' is set to a value higher than the highest cell value in 'CHM'")
      if (minHeight > CHM.rng["min"]) {
        CHM[CHM < minHeight] <- NA
        CHM.rng["min"] <- minHeight
      }
    }
    if (verbose)
      radii <- seq(floor(winFun(CHM.rng["min"])), ceiling(winFun(CHM.rng["max"])),
                   by = roundRes[1])
    radii <- radii[radii >= roundRes[1]]
    if (length(radii) == 0) {
      warning("The maximum window radius computed with 'winFun' is smaller than the CHM's resolution",
              "\nA 3x3 cell search window will be uniformly applied",
              "\nUse a higher resolution 'CHM' or adjust 'winFun' to produce wider dynamic windows")
      radii <- roundRes[1]
    }
    maxDimension <- (max(radii)/roundRes[1]) * 2 + 1
    if (!is.null(maxWinDiameter) && maxDimension > maxWinDiameter) {
      stop("Input function for 'winFun' yields a window diameter of ",
           maxDimension, " cells, which is wider than the maximum allowable value in 'maxWinDiameter'.",
           "\nChange the 'winFun' function or set 'maxWinDiameter' to a higher value (or to NULL).")
    }
    windows <- lapply(radii, function(radius) {
      win.mat <- raster::focalWeight(raster::raster(resolution = roundRes),
                                     radius, type = "circle")
      if (nrow(win.mat) == 3 && minWinNeib == "queen")
        win.mat[] <- 1
      win.pad <- raster::extend(raster::raster(win.mat), (maxDimension -
                                                            ncol(win.mat))/2, value = 0)
      win.vec <- as.vector(win.pad != 0)
      return(win.vec)
    })
    names(windows) <- radii
    .vwMax <- function(x, ...) {
      centralValue <- x[length(x)/2 + 0.5]
      if (is.na(centralValue)) {
        return(NA)
      }
      else {
        radius <- winFun(centralValue)
        window <- windows[[which.min(abs(as.numeric(names(windows)) -
                                           radius))]]
        return(if (max(x[window], na.rm = TRUE) == centralValue) 1 else 0)
      }
    }
    lm.pts <- raster::rasterToPoints(raster::focal(CHM, w = matrix(1,
                                                                   maxDimension, maxDimension), fun = .vwMax, pad = TRUE,
                                                   padValue = NA), fun = function(x) x == 1, spatial = TRUE)
    lm.pts[["height"]] <- raster::extract(CHM, lm.pts)
    lm.pts[["winRadius"]] <- winFun(lm.pts[["height"]])
    lm.pts[["treeID"]] <- 1:length(lm.pts)
    return(lm.pts)
  } ############################################################################

  # function ForestTools mcws cleaned from cat code #############################
  mcws_clean <- function (treetops, CHM, minHeight = 0, format = "raster", OSGeoPath = NULL,
                                verbose = FALSE)
  {
    if (verbose)
      if (!toupper(format) %in% c("RASTER", "POLYGONS", "POLYGON",
                                  "POLY")) {
        stop("'format' must be set to either 'raster' or 'polygons'")
      }
    if (verbose)
      CHM.max <- suppressWarnings(raster::cellStats(CHM, "max"))
    if (is.infinite(CHM.max)) {
      stop("Input CHM does not contain any usable values.")
    }
    if (minHeight > CHM.max) {
      stop("'minHeight' is set higher than the highest cell value in 'CHM'")
    }
    raster::crs(CHM) <- raster::crs(treetops)
    treetopsHgts <- raster::extract(CHM, treetops)
    treetops <- treetops[!is.na(treetopsHgts) & treetopsHgts >=
                           minHeight, ]
    if (length(treetops) == 0) {
      stop("No usable treetops. Treetops are either outside of CHM's extent, or are located elow the 'minHeight' value")
    }
    if (!"treeID" %in% names(treetops)) {
      warning("No 'treeID' found for input treetops. New 'treeID' identifiers will be added to segments")
      treetops[["treeID"]] <- 1:length(treetops)
    }
    else {
      if (any(treetops[["treeID"]] == 0))
        stop("'treeID' cannot be equal to 0")
      if (any(duplicated(treetops[["treeID"]])))
        warning("Duplicate 'treeID' identifiers detected")
    }
    if (verbose)
      CHM.mask <- is.na(CHM) | CHM < minHeight
    CHM[CHM.mask] <- 0
    if (verbose)
      ttops.ras <- raster::rasterize(treetops, CHM, "treeID", background = 0)
    CHM.img <- imager::as.cimg(raster::as.matrix(CHM))
    ttops.img <- imager::as.cimg(raster::as.matrix(ttops.ras))
    if (verbose)
      ws.img <- imager::watershed(ttops.img, CHM.img)
    ws.ras <- raster::raster(vals = ws.img[, , 1, 1], nrows = nrow(CHM),
                             ncols = ncol(CHM), ext = raster::extent(CHM), crs = raster::crs(CHM))
    ws.ras[CHM.mask] <- NA
    if (toupper(format) %in% c("POLYGONS", "POLYGON", "POLY")) {
      if (verbose)
        if (is.null(OSGeoPath)) {
          polys <- raster::rasterToPolygons(ws.ras)
          polys <- rgeos::gUnaryUnion(polys, id = polys[["layer"]])
          polys <- sp::disaggregate(polys)
        }
      else {
        polys <- APfun::APpolygonize(ws.ras, OSGeoPath = OSGeoPath)
      }
      if (verbose)
        polys.over <- sp::over(polys, treetops)
      polys.out <- sp::SpatialPolygonsDataFrame(polys, subset(polys.over,
                                                              select = which(names(polys.over) != "treeID")))
      polys.out <- polys.out[match(treetops[["treeID"]], polys.over[,
                                                                    "treeID"]), ]
      if (verbose)
        if ("crownArea" %in% names(polys.out)) {
          i <- 1
          while (paste0("crownArea", i) %in% names(polys.out)) i <- i +
              1
          crownArea <- paste0("crownArea", i)
          warning("Input data already has a 'crownArea' field. Writing new crown area values to the 'crownArea",
                  i, "' field")
        }
      else {
        crownArea <- "crownArea"
      }
      polys.out[[crownArea]] <- rgeos::gArea(polys.out, byid = TRUE)
      if (verbose)
        return(polys.out)
    }
    else {
      if (verbose)
        cat("..Returning crown outlines as a raster\n\nFinished at:",
            format(Sys.time(), "%Y-%m-%d, %X"), "\n")
      return(ws.ras)
    }
  }#############################################################################

  # setup UAVRST chmseg
  chmseg_FT <- function(treepos = NULL,
                        chm = NULL,
                        minTreeAlt = 2,
                        format = "polygons",
                        winRadius = 1.5,
                        verbose = FALSE) {

    if (class(treepos) %in% c("RasterLayer", "RasterStack", "RasterBrick")) {
      # add projection of chm TODO
      pr<-raster::crs(raster::projection(raster::raster(chm)))
      raster::projection(treepos) <- pr
      treepos <- raster::rasterToPoints(treepos,spatial = TRUE)
      # reformat it to the needs of mcws
      treepos@data$layer <- 1
      treepos@data$winRadius <- 1.5
      treepos@data$treeID <- seq(1:nrow(treepos@data))
      names(treepos)<-c("height","treeID","winRadius","layer")
    }
    # Crown segmentation
    crownsFT <- mcws_clean(treetops = treepos,
                                  CHM = chm,
                                  format = format,
                                  minHeight = minTreeAlt,
                                  verbose = verbose)
    return(crownsFT)
  } ############################################################################

  # TreeSeg

  # check inputs
  if(CHMfilter %% 2 == 0){
    stop("selected 'filter' sizes contain even values (use odd values only)")
  }
  # check filter
  if(CHMfilter!=1){
    if(silent==FALSE){
    cat(paste0("### CENITH computes ",CHMfilter,"*",CHMfilter," sum filter for chm ###",sep = "\n"))
    }
    chm <-raster::focal(chm,w=matrix(1/(CHMfilter*CHMfilter),nrow=CHMfilter,ncol=CHMfilter),fun=sum,na.rm=TRUE)
  }

  if(silent==FALSE){
  cat(paste0("### CENITH starts segmentation ###",sep = "\n"))
  }
  # compute Treepositions
  tpos = vwf_clean(chm,
                   winFun = function(x){x * a + b},
                   minHeight = h,
                   verbose = TRUE)

  # compute segments
  seg <- chmseg_FT(chm = chm,
                           treepos = tpos,
                           format = "polygons",
                           minTreeAlt = h,
                           verbose = TRUE)
  if(silent==FALSE){
    cat(paste0("detected ",length(tpos), " trees ###",sep = "\n"))
  }

  # clip min and max
  if(silent==FALSE){
    cat(paste0("### CENITH starts cropping to MIN and MAX ###",sep = "\n"))
  }
  seg_min <- seg[seg$crownArea>MIN,]
  seg_max <- seg_min[seg_min$crownArea<MAX,]
  seg <- seg_max
  if(length(seg)==0){
    stop(" !!! after clipping to MIN and MAX no polygons are returned, setting results to NA",sep = "\n")}

  if(silent==FALSE){
    cat(paste0("clipped: ",length(tpos)-length(seg)," polygons. ", length(seg)," trees remaining",sep = "\n"))
    cat(paste0("### CENITH finished segmentation ###",sep = "\n"))

  }
  return(seg)
}#end of main function


R/CENITH - package.R
#' CENITH Tree Segmentation Tool
#'
#' @name LEGION - package
#' @docType package
#' @title CENITH Tree Segmentation Tool
#' @description In general the \code{CENITH} package is a wrapper for the \code{ForestTools} package (Andrew Plowright and Jean-Romain Roussel,2020).
#' \code{CENITH} is designed to test for best fitting parameters to performe a Tree Segmentation. Furher it provides a cross-validation for tree segmentation.
#'
#' @note For a more independent and stable package the source code of some fucntions provided by \code{ForestTools} is used instead of the respective function.
#' Possible changes of \code{ForestTools} in the future will NOT change the workflow of \code{CENITH}
#'
#' @author Andreas Schönberg
#' @import raster
#' @import ForestTools
#' @keywords package
NULL
#' @docType data
#' @name lau_AOI_chm - data
#' @title A canopy height model (chm) of the AOI for the tutorial
#' @description  chm of the AOI in the Lautaret valley in the French Alps. Resolution 0.15 m.
#' @format \code{"raster::raster"}
NULL
#' @docType data
#' @name lau_AOI_rgb - data
#' @title An RGB Raster image of the AOI for the tutorial
#' @description RGB Raster of the AOI in the Lautaret valley in the French Alps. Resolution 0.15 m.
#' @format \code{"raster::raster"}
NULL
#' @docType data
#' @name lau_chm - data
#' @title A canopy height model (chm)
#' @description CHM of trees in site 1 of 3 in the Lautaret valley in the French Alps. Resolution 0.15 m.
#' @format \code{"raster::raster"}
NULL
#' @docType data
#' @name lau_chm2 - data
#' @title A canopy height model (chm)
#' @description CHM of trees in site 2 of 3 in the Lautaret valley in the French Alps. Resolution 0.15 m.
#' @format \code{"raster::raster"}
NULL
#' @docType data
#' @name lau_chm3 - data
#' @title A canopy height model (chm)
#' @description CHM of trees in site 3 of 3 in the Lautaret valley in the French Alps. Resolution 0.15 meter.
#' @format \code{"raster::raster"}
NULL
#' @docType data
#' @name lau_vp - data
#' @title Supervised computed Treepositions
#' @description Estimated position of trees for the first example chm in the Lautaret Valley. It is used to validate the segmentation.
#' @format \code{"rgdal::readOGR"}
NULL
#' @docType data
#' @name lau_vp_site2 - data
#' @title Supervised computed Treepositions
#' @description Estimated position of trees for the second example chm in the Lautaret Valley. It is used to validate the segmentation.
#' @format \code{"rgdal::readOGR"}
NULL
#' @docType data
#' @name lau_vp_site3 - data
#' @title Supervised computed Treepositions
#' @description Estimated position of trees for the third example chm in the Lautaret Valley. It is used to validate the segmentation.
#' @format \code{"rgdal::readOGR"}
NULL



R/BestSegVal.R
#' Test for best performing parameters for TreeSeg TreeSegmentation using Validation Treepositions
#' @description Iterates over a,b,h,MIN and supports iterating over filtered chms. Uses supervised computed Treepositions to validate best fitting values for a,b, height, MIN and CHM filters to detect trees.
#' @param chm raster -  Canopy Height Model RasterLayer derived form LiDAR data
#' @param a numeric - single value, combination of values or sequence for MovingWindow
#' @param b numeric - single value, combination of values or sequence for MovingWindow
#' @param h numeric - single value, combination of values or a sequence for the maximum height of trees (in meter) to detect trees.
#' @param vp shp - PointLayer with estimated Positions of trees (see details).
#' @param MIN numeric - single value, combination of values or a sequence of minimum area for crowns. Smaller polygons are cropped Default= 0
#' @param MAX numeric - the maximum area for crowns. Larger polygons are cropped. Default=1000
#' @param filter numeric - single value, combination of values or a sequence for filtersize, uses a sum filter on the chm with a MovingWindow of (x*x), which must be odd., default= 1 (no filter.)
#' @param skipCheck for development - bolean - if TRUE skips checking input data and estimation of ETA.

#' @return returns a dataframe with several validation scores (see details).
#' @details
#' * 'start and go to sleep' The function has implemented error catching. If an iteration would cause a critical stop error, the loops continue. Returns NA for corrupted iterations.
#' * ETA: before starting the iterations, up to 3 random combinations are tested and the processing time is recorded. The processing time is multiplied with the total amount of iterations to calculate the ETA.
#' * Input for a, b, h, MIN and filter supports - numeric, single and combination: c(), or sequence: seq() to iterate over: (e.g a=0.5 ; a=c(0.3,0.5) ; a=seq(0.1,0.9,0.05).
#' * filter - it could be helpful to filter the raw chm to 'smooth' small peaks and 'holes' in crowns. Helps to suppress the segmentation of many tiny objects (like stones) and leads to polygons with less 'holes'.
#' * vp - validation point: supervised computed pointlayer. Use a GIS like Qgis to set points where Trees are estimated (see Tutorial).
#' * if 'skipCheck' = TRUE the input checking is skipped and the iterations start directly.
#' * result table
#'    + a, b, height, MIN, chm - the used parameters
#'    + total_seg - total computed segments
#'    + hit/vp - amount of segments computed for trees marked with validation points
#'    + under - undersegmnetation in absolut values, amount of segments which contain more than one validation point.
#'    + over - oversegmentation in absolut values, amount of segments without validation point
#'    + area - the total area of segments
#'    + hitrate - percentage of segments which contain exactly one validation point in relation to the total validation points.
#'    + underrate - undersegmentation in relation to the total segments.
#'    + overrate - oversegmentation in relation to the total segments.
#'    + Seg_qualy - quality of segmentation:  hitrate @ combined over- and undersegmentations. A segment with more than one validation points is estimated to be better than segments without validation points. Therefore the 'miss' value is (over+(2*under))/2.
#' @author Andreas Schönberg
#' @note The 'brute force' approach to iterate over many parameters may result in very long time to finish. Preselected smaller samples may be more efficient (see Tutorial).
#' @examples
#' # load packages
#' require(CENITH)
#' require(mapview)
#' require(raster)
#' require(rgdal)
#' ### NOTE: this example is used to show the usage of 'BestSegVal'. It is NOT used to show a workflow for best Results (see 'Tutorial' vignette for workflow strategies)
#' ### further NOTE: to reduce time usage for the example only small iteartions are used to get an overlook for the functionallities.
#' # load data
#' chm <- raster::raster(system.file("extdata","lau_chm.tif",package = "CENITH"))
#' vp <- rgdal::readOGR(system.file("extdata","lau_vp.shp",package = "CENITH"))
#' # handle CRS string
#' crs(vp)<-crs(chm)
#' # take a look at the data (chm with estimated Treeposition)
#' mapview(chm)+vp
#' # start BestSegVal with parameters for Moving window and minimum height
#' x <- BestSegVal(chm,a=c(0.3,0.5),b=c(0.5,0.7),h=c(0.1,1),vp = vp,filter=c(1,3))
#' # check for best hitrate
#' maxrow <- x[which.max(x$hitrate),] # search max vale but rturn only 1 value
#' maxhit <- maxrow$hitrate
#' x[which(x$hitrate==maxhit),]
#' # visualise best combination (due to Segment Quality)
#' y <- TreeSeg(chm,0.3,0.7,0.1,CHMfilter = 3)
#' # show result segments with validation points and chm
#' mapview(chm)+vp+y
#' # BestSegVal based on this results keep a,b and filter 3 with MIN values to reduce oversegmentation and different heights.
#' z <- BestSegVal(chm,a=0.3,b=0.7,h=c(0.1,0.2,0.5,1),MIN=c(10,100),vp = vp,filter=3)
#' ### Note that row 1 and 2 lead to much lesser oversegmnetation just differ in area.
#' # visualize
#' v <- TreeSeg(chm,0.3,0.7,0.1,MIN=10,CHMfilter = 3)
#' mapview(chm)+vp+v

#' @export BestSegVal
#' @aliases BestSegVal


BestSegVal<- function(chm,a,b,h,vp,MIN=0,MAX=1000,skipCheck=FALSE,filter=1){
  cat(paste0("### CENITH starts Best Parameter Validation ###",sep = "\n"))
  if(skipCheck==FALSE){
    cat(paste0("checking input ...",sep = "\n"))
    #check for wrong sizes input
    if(any(filter %% 2 == 0)){
      stop("selected 'filter' sizes contain even values (use odd values only)")
    }


    #checking projection
    if(compareCRS(chm,vp)==FALSE){
      stop("unequal CRS for 'chm' and 'vp' detected !")
    }
    # check input height is lesser than max height in chm
    if(cellStats(chm,max)<=max(h)){
      stop("input height 'h' values are higher than the highest cell value of teh CHM")
    }

    # estimate time remaining, trying up to tree iterations, if all fail, skipping time estimation

    cat(paste0("calculating estimate time to finish ...",sep = "\n"))
    cat(" ",sep = "\n")

    start <- Sys.time()

    ETA1 <-try(TreeSeg(chm,a=sample(a,1),b=sample(b,1),h=sample(h,1),MIN=sample(MIN,1),MAX,CHMfilter=1,silent=TRUE),silent = T)
    stop  <- Sys.time()
    if(class(ETA1)=="try-error"){
      start <- Sys.time()
      ETA2 <-try(TreeSeg(chm,a=sample(a,1),b=sample(b,1),h=sample(h,1),MIN=sample(MIN,1),MAX,CHMfilter=1,silent=TRUE),silent = T)
      stop  <- Sys.time()
      if(class(ETA2)=="try-error"){
        start <- Sys.time()
        ETA3 <-try(TreeSeg(chm,a=sample(a,1),b=sample(b,1),h=sample(h,1),MIN=sample(MIN,1),MAX,CHMfilter=1,silent=TRUE),silent = T)
        stop  <- Sys.time()
        if(class(ETA3)=="try-error"){
          cat("unable to calculate ETA, tried 3 random iterations, skipping.",sep = "\n")
          cat("#----------------------------------------#",sep = "\n")
          stop <- start}
      }
    }



  if(start!=stop){
    # amount of iterations
    na <- length(a)
    nb <- length(b)
    nh <- length(h)
    nm <- length(MIN)
    nc <- length(filter)
    niter <- nh*nb*na*nm*nc

    # test time for 1 iteration
    timedif <-difftime(stop,start,units = "min")
    # predict estimated time by multiply timedif with n iterations

    eta <- timedif*niter
    eta <-1.2*eta # add 20% because the iteration wrapper needs more time.
    cat(paste("requires",niter,"iterations @",round(timedif,4),units(timedif),"per iteration"),sep="\n")
    cat(paste("estimated",round(eta,4),units(timedif),"to finish"),sep = "\n")
    cat("#----------------------------------------#",sep = "\n")
  }

    # warning / wait for input if niter is too big
    #if(niter>50&&niter<100){
    #  cat("",sep = "\n")
    #  cat(paste0("WARNING over 50 iterations detected this may take long time to finsih, maybe abort and start smaller trys (see details)",sep = "\n"))
    #}
    #if(niter>=100){
    #  abort <- readline(prompt="More than 100 iterations detected! Are you sure you want to continue? 1:YES, 2:NO, plz abort ")
    #  if(abort==2){
    #    stop("aborded on users desicion, too much iterations detected")
    #  }
    #}
  }# end skip check

  # setup function to iterate over b ############################################################
  loop_b <- function(chm,a,b,h,vp,MIN,MAX){


    # create dataframe to save informations
    result <- data.frame(matrix(nrow = length(b), ncol = 8)) # ncol = n information stored

    # loop to iterate on varibale
    for (j in seq(1:length(b))){

      cat(paste0("iterating over b ",as.factor(j),"/",as.factor(length(b))," ",sep = "\n"))
      # TreeSeg
      seg <- try(TreeSeg(chm=chm,a,b[j],h,MIN,MAX,silent = TRUE,CHMfilter=1),silent=T)
      # handle error catch
      if(class(seg)=="try-error"){
        cat(paste0(" !!! iteration a=",a," b=",b[j]," h=",h," leads to an error, setting to 'NA' in results ",sep="\n"))
        result[j, 1] <- a
        result[j, 2] <- b[j]
        result[j, 3] <- h
        #absolut results
        result[j, 4] <- NA
        result[j, 5] <- NA
        result[j, 6] <- NA
        result[j, 7] <- NA
        result[j, 8] <- NA
        # rates
        result[j, 9]  <- NA
        result[j, 10] <- NA
        result[j, 11] <- NA
        result[j, 12] <- NA
        result[j, 13] <- MIN
        result[j, 14] <- names(chm)


      } # if error
      else{


        ### save results in dataframe################################

        # get stats of overlapping
        stat <- ForestTools::sp_summarise(vp, seg) # compute data points in polygons
        stat[is.na(stat$TreeCount)] <- 0 # na to 0

        # get n trees in poygons
        TCO <- sum(stat$TreeCount<1) # amount polygon without any tree (miss) <- Oversegmented
        TC1 <- sum(stat$TreeCount==1) # amount polygon with exact 1 tree (hit)
        TCU <- sum(stat$TreeCount>1) # amount polygons with more than 1 tree (miss) <- Undersegmented
        TC2 <- sum(stat$TreeCount==2) # amount polygon with 2 Trees (miss)
        TC3 <- sum(stat$TreeCount==3) # amount polygon with 3 Trees (miss)
        TCX <- sum(stat$TreeCount>3)  # amount polygon more tha 3 Trees (miss)

        # calculate validation scores
        # absolute
        hit = paste(TC1,"/",length(vp))

        # rates
        hitrate = TC1/length(vp)        # hits / seg amount of hits in relation to nseg
        over = TCO/length(stat$TreeCount)           # oversegmented, Segnents with no VP
        under = TCU/length(stat$TreeCount) # undersegmented, Segments with more than one tree

        # quality value calculation
        miss <- (over+(2*under))/2
        segQy <- paste0(round(hitrate,2)," @ ",round(miss,2))
        segQy
        # additional informations
        tseg = length(seg)
        area =  sum(seg$crownArea)#

        # write out informations in dataframe
        result[j, 1] <- a
        result[j, 2] <- b[j]
        result[j, 3] <- h
        #absolut results
        result[j, 4] <- tseg
        result[j, 5] <- hit
        result[j, 6] <- TCU
        result[j, 7] <- TCO
        result[j, 8] <- area
        # rates
        result[j, 9] <- hitrate
        result[j, 10] <- under
        result[j, 11] <- over
        result[j, 12] <- segQy
        result[j, 13] <- MIN
        result[j, 14] <- names(chm)
      } # end of more than null polygons


    }# end loop for b
    # output
    return(result)
  }# end of iteration over b #####################################################

  # setup function to iterate over a ############################################################
  loop_a <- function(chm,a,b,h,vp,MIN,MAX){
    for (i in seq(1:length(a))){
      cat("",sep = "\n")
      cat(paste0("### CENITH starts iterating over a ",as.factor(i),"/",as.factor(length(a))," ",sep = "\n"))
      if(i==1){
        res <-loop_b(chm,a[i],b,h,vp,MIN,MAX)

      }    else {
        res2 <-loop_b(chm,a[i],b,h,vp,MIN,MAX)
        res= rbind(res,res2)
      }


    }# end loop for a

    return(res)
  }# end of iteration a



  # iteration h
  loop_h <- function(chm,a,b,h,vp,MIN,MAX){
    for (c in seq(1:length(h))){

      cat(paste0("### CENITH starts iterating over h ",as.factor(c),"/",as.factor(length(h))," ###",sep = "\n"))
      if(c==1){
        res <-loop_a(chm,a,b,h[c],vp,MIN,MAX)
      }    else {
        res2 <-loop_a(chm,a,b,h[c],vp,MIN,MAX)
        res= rbind(res,res2)}

    }# end loop h
    return(res)
  }# end iteration h

  # iteration h
  loop_MIN <- function(chm,a,b,h,vp,MIN,MAX){
    for (k in seq(1:length(MIN))){

      cat(paste0("### CENITH starts iterating over MIN ",as.factor(k),"/",as.factor(length(MIN))," ###",sep = "\n"))
      if(k==1){
        res <-loop_h(chm,a,b,h,vp,MIN[k],MAX)
      }    else {
        res2 <-loop_h(chm,a,b,h,vp,MIN[k],MAX)
        res= rbind(res,res2)}

    }# end loop h
    return(res)
  }# end iteration h





  # BestSeg Core


  for(l in seq(1:length(filter))){
    cat("",sep = "\n")
    cat(paste0("### CENITH starts iterating over CHM ",as.factor(l)),"/",(length(filter))," ###")
    cat("",sep = "\n")
    if(l==1){
      #for no filtering with 1
      if(filter[l]==1){
        res <-loop_MIN(chm,a,b,h,vp,MIN,MAX)
      } else {
        chmf <- raster::focal(chm,w=matrix(1/(filter[l]*filter[l]),nrow=filter[l],ncol=filter[l]),fun=sum,na.rm=TRUE)
        names(chmf) <- paste0(names(chm),"_",as.factor(filter[l]))
        res <-loop_MIN(chmf,a,b,h,vp,MIN,MAX)
      }
    }    else {
      chmf <- raster::focal(chm,w=matrix(1/(filter[l]*filter[l]),nrow=filter[l],ncol=filter[l]),fun=sum,na.rm=TRUE)
      names(chmf) <- paste0(names(chm),"_",as.factor(filter[l]))
      res2 <-loop_MIN(chmf,a,b,h,vp,MIN,MAX)
      res= rbind(res,res2)}

  }# end core function


  # output
  cat("",sep = "\n")
  cat(paste0("### CENITH finished Best Parameters Validation ###"),sep = "\n")

  names(res)<- c("a","b","height","total_seg","hit/vp","under","over","area","hitrate","underrate","overrate","Seg_qualy","MIN","chm")
  #names(res)<- c("a","b","height","hit","tp/vp_rate","tpos/vp","miss","area","empty")
  return(res)
}# end core fucntion
